[blog] 12-20 release (#874)
gitpod-io/website: src/blog/release-12-20.md

---
date: Thu Dec 07 2020 11:49:24 GMT+0000 (UTC)
author: svenefftinge
title: Root, Docker and VS Code
subtitle: New Release
image: /globe.png
teaserImage: /globe.png
---

The team has been focussing on shipping some major new features, which mark important milestones on our mission to **eliminate all friction around maintaining dev environments**. 

> We are rolling out new features in [preview mode](/docs/feature-preview) first. So make sure to switch __Feature Preview__ on in [your settings](https://gitpod.io/settings/) if you want to try them out.

 - [Root Access](#root-access)
 - [Docker Support](#docker)
 - [VS Code and other IDEs](#vs-code)
 - [Connect to self-managed GitLab](#self-managed-gitlab)

## Root Access 💪{#root-access}

> Gitpod workspaces run as containers which make them incredibly fast and efficient. The main disadvantage in comparison to VMs is that isolating workspaces from each other is not trivial. Until now we had to restrict users so that they didn’t get any relevant privileges in their workspaces. As a result you couldn’t install additional tools using your favorite package manager or do things like running docker or docker-compose. Obviously this has made some things more difficult than they should be. 

Reviewer: ```suggestion
Gitpod workspaces run as containers which make them incredibly fast and efficient. Compared to VMs, isolating workspace containers from each other is not easy. Due to this difficult isolation, Gitpod operated a fairly restricted set of privileges in within its workspaces. This meant you couldn’t install additional tools using your favourite package manager (think `sudo apt-get install ...`) or run Docker/`docker-compose`. Clearly, this has made some things more difficult than they should be. 
```

> After investigating different options such as gvisor or running VMs, we’ve settled on a solution that isolates through namespace mappings. Chris, our Chief Architect and author of the isolation mechanics, has explained how it works in detail in this webinar. Long story short, you can now run commands using `sudo`. So for instance, you can test-install additional packages using apt-get, before adding those commands to your workspace’s dockerfile (you should really put it there eventually).

Reviewer: ```suggestion
After investigating different options such as [gVisor](https://github.com/google/gvisor) or [virtual machines](https://katacontainers.io/), we’ve settled on a solution that isolates through user namespaces. Chris, our Chief Architect and author of the isolation mechanics, has explained how it works in detail in [this webinar](https://www.youtube.com/watch?v=l4I2TVAnBuw). Long story short, you can now run commands using `sudo`. So for instance, you can test-install additional packages using `apt-get`, before adding those commands to your workspace’s Dockerfile (where setup commands should still go eventually).
```

`youtube: 5JUfzet1NNg`

## Docker 🐳{#docker}

> With the new privileges you can now also run build docker images start containers within your workspace. Gitpod’s default image (workspace-full) comes equipped with docker now, so all you need to do is run `sudo docker-up` and wait until the service is listening.

Reviewer: ```suggestion
With the new privileges you can now also run and build Docker images to start containers within your workspace. Gitpod’s default image (workspace-full) comes equipped with Docker now, so all you need to do is run `sudo docker-up` and wait until the service is listening.
```

> Now start another terminal and use the docker cli as usual. Here is an example repo, that starts a small hello world example.

Reviewer: ```suggestion
Now start another terminal and use the Docker CLI as usual. Here is an example repo, that starts a small hello world example.
```

`youtube: https://youtu.be/HnEXAO2P93Q`

> > Warning: This is in feature preview for a reason, as there are some things we still have to get straight. For instance, process ids are not properly mapped which might cause some weird errors. Please provide feedback in the [community](https://community.gitpod.io) or the [issue tracker](https://www.github.com/gitpod-io/gitpod/issues).

Reviewer: ```suggestion
> Warning: This is in feature preview for a reason, and there are a few things we still have to sort out. For instance, within Docker containers process ids are not properly mapped which might cause some weird errors. Please provide feedback in the [community](https://community.gitpod.io) or the [issue tracker](https://www.github.com/gitpod-io/gitpod/issues).
```

## VS Code and Other IDEs 🥂{#vs-code}

> At Gitpod we want to help automate your dev environments, so that you are always ready-to-code. But we don’t think you should make any compromises in terms of what tools you use. Instead we make sure you get all your favorite tools readily configured and initialized for any project and task. Until today you had no choice but had to use Theia, which is a great vendor-neutral IDE, that leverages a lot of VS Code’s tech and provides a very similar UX. Still there are some differences and since VS Code has managed to change its architecture now to properly run in browsers as well, we are supporting proper VS Code as an alternative.

Reviewer: ```suggestion
At Gitpod we want to help automate your dev environments, so that you are always ready-to-code. We believe you should not have to compromise in terms of what tools you use. Instead we make sure you get all your favourite tools readily configured and initialised for any project and task. Until today you had no choice but to use Theia, which is a great vendor-neutral IDE that leverages a lot of VS Code’s tech and provides a very similar UX. Still there are some differences in behaviour and compatibility. Now that VS Code has changed its architecture to run in browsers as well, we are supporting VS Code, too.
```

> You can even bring your own web IDE though a docker image, to run other IDEs such as all the Jetbrains IDEs (using projector), Jupyter Notebook, or Jupyter Lab. This feature is however currently still in a private beta, so you need to reach out in order to get access to that.

Reviewer: ```suggestion
You can even bring your own web IDE though a Docker image. This enables Gitpod to run other IDEs such as all the Jetbrains IDEs (using [projector](https://github.com/JetBrains/projector-server/blob/master/docs/Projector.md)), Jupyter Notebook, or Jupyter Lab. This feature is however currently still in a private beta, so you need to reach out in order to get access to that.
```

`youtube: 5hArIRPZuBI`

> Warning: This is an early preview of the functionality and we haven't fully implemented all necessary features, yet. For instance, user settings are not persisted among sessions. Also while you can install extensions, they are also not persisted and extension settings from `.gitpod.yml` are currently ignored. We'll follow up with these capabilities ASAP.

> Please provide feedback in the [community](https://community.gitpod.io) or the [issue tracker](https://www.github.com/gitpod-io/gitpod/issues)

We are also shipping an SSH mode in Q1/21, that will let you connect to Gitpod using your favorite desktop IDE’s remote mode.

## Connect to Self-managed GitLab ❤️{#self-managed-gitlab}

> In case you don’t want to host and operate your own Gitpod instance somewhere, your can now [hook up your self-managed GitLab application with Gitpod.io](/docs/gitlab-integration#oauth-application). To do so, go to the settings and scroll down to the “Git Provider Integrations” section. Add your instance and follow the instructions. 

Reviewer: ```suggestion
You can now [connect your self-hosted GitLab application with gitpod.io](/docs/gitlab-integration#oauth-application). This way you don't have to host and operate your own Gitpod installation, just to make it work with your GitLab installation. To make this connection, go to the settings and scroll down to the “Git Provider Integrations” section. Add your instance and follow the instructions. 
```

> When you are done, your entire team can open dev environments by prefixing the projects and merge requests URLs on the self-managed instance with `gitpod.io#` or simply use the Gitpod button provided by GitLab (make sure it is enabled).

Reviewer: ```suggestion
After making this connection, your entire team can open dev environments by prefixing the projects and merge requests URLs on the self-managed instance with `gitpod.io#` or simply use the Gitpod button provided by GitLab (make sure it is enabled).
```

![Gitpod loves GitLab](/teaser-gitlab-gitpod.jpg)

## Other Updates

This release comes with many other bugfixes and features. Also we’ve updated to a newer version of Theia. TODO...

## Next Steps ⏩

> There will be a Gitpod Self-Hosted release before Christmas that includes the features listed here as well as some much additional enhancements and bugfixes for installing and running Gitpod on different infrastructures. Furthermore we are completing the features in preview, work on some important updates around prebuilds and are kicking off a redesign of the dashboard. The future is bright 🌅.

Reviewer: ```suggestion
There will be a Gitpod Self-Hosted release before Christmas that includes the features listed here as well as many additional enhancements and bugfixes for installing and running Gitpod on different infrastructures. 
In the near future, we will complete the features in preview, work on some important updates around prebuilds, support SSH access to workspaces, and are kicking off a redesign of the dashboard. The future is bright 🌅.
```
